### PR TITLE
First PHP CNB implementation

### PR DESCRIPTION
**Pull request by dzuelke**
Still to do, coming in separate PRs:
- some trivial validation (see `validate` in `php_project.rs`)
- output filtering for `composer` and other invocations
- some refactoring of `PlatformLayer` in conjunction with corresponding changes to the Composer Installer Plugin

Pretty much all `TODO` or `FIXME` comments are either in reference to the above items, or longer-term notes.

Related: https://github.com/heroku/heroku-buildpack-php/pull/new/cnb-installer

### CODE REVIEW

File: buildpacks/php/src/main.rs
Comment (edmorley) [inline buildpacks/php/src/main.rs:1]: Running `cargo clippy --all-targets` on this branch locally with Rust 1.70, I get 62 warnings across the two crates, of which 38 are auto-correctable using the `--fix` mode. Just checking that:
- you know about `--fix` (in case it saves you some time)
- you know about the whole IDE `check` vs `clippy` command config thing (IDEs tend to default to running the `check` command rather than the `clippy` command, but for VSCode at least can be configured to run the latter, if avoiding surprises in CI is preferred)

**Review by edmorley — COMMENTED** [review]
Awesome to see the PHP CNB coming together!

I know the PR is marked as WIP, but I spotted a couple of things whilst I was reading through and thought I'd leave comments rather than end up having to remember them when it comes to the real review :-)

**Review by Malax — COMMENTED** [review]
First pass over the code, ignoring obvious issues like remaining TODO comments, `dbg!` and similar. Structure is good, no need to dramatically shift things around, left a couple of nitpicks and ideas here and there.

File: buildpacks/php/src/main.rs
Comment (dzuelke) [reply buildpacks/php/src/main.rs:53]: Yeah... this has bothered me, too, but not sure yet how to do it cleanly. How about returning it via the layer metadata, would you say that is completely wrong?

File: buildpacks/php/src/utils.rs
Comment (dzuelke) [reply buildpacks/php/src/utils.rs:53]: I'd love to, but Clippy always complains about these and auto-fixes them back into `if`s, e.g. here: https://github.com/heroku/buildpacks-php/pull/2/commits/4587f6a555506e44dcf85640562563bba629833b#diff-887bbd9e2d55e3d4905bb3ffd054a8229c67a52aa4cfba2a5d9f3ba5fb6f9b9cL565-L568

File: buildpacks/php/src/layers/bootstrap.rs
Comment (dzuelke) [reply buildpacks/php/src/layers/bootstrap.rs:36]: The bootstrap layer points to branch `cnb-installer` of the classic buildpack right now, but will switch to a tag (and, a bit longer-term, somehow incorporate it into the CNB build, or mirror to S3, so we do not have to download something off GitHub during bootstrapping).

File: buildpacks/php/src/layers/composer_env.rs
Comment (dzuelke) [reply buildpacks/php/src/layers/composer_env.rs:35]: Yeah, so... different buildpacks have it in different places, and for e.g. the bootstrap layer, it has to announce that it's restoring from cache... but in other layers, I do not want to log the header in the layer, since I use it multiple times for different purposes (platform for platform and webservers install)

File: buildpacks/php/src/main.rs
Comment (dzuelke) [reply buildpacks/php/src/main.rs:116]: I am simply following @edmorley's shining example :p

File: buildpacks/php/src/utils.rs
Comment (dzuelke) [reply buildpacks/php/src/utils.rs:53]: (not this specific one, but similar statements)

File: buildpacks/php/src/platform/generator.rs
Comment (dzuelke) [reply buildpacks/php/src/platform/generator.rs:186]: Yup, leftover from when that used to have more packages there and the ton of `.to.string()`s were hurting my eyes ;)

File: buildpacks/php/src/platform/generator.rs
Comment (dzuelke) [reply buildpacks/php/src/platform/generator.rs:160]: As a first improvement, how about named capture groups? ;)

**Review by edmorley — APPROVED** [review]
Awesome! 😄 🎉

File: buildpacks/php/src/layers/composer_cache.rs
Comment (dzuelke) [reply buildpacks/php/src/layers/composer_cache.rs:33]: good call, will check

File: buildpacks/php/src/layers/composer_cache.rs
Comment (dzuelke) [reply buildpacks/php/src/layers/composer_cache.rs:44]: downloads only!

File: buildpacks/php/src/main.rs
Comment (dzuelke) [reply buildpacks/php/src/main.rs:36]: ugh yeah this is IntelliJ appending its imports to the last `use` it sees

File: buildpacks/php/src/layers/composer_cache.rs
Comment (dzuelke) [reply buildpacks/php/src/layers/composer_cache.rs:33]: subdirs: https://getcomposer.org/doc/06-config.md#cache-dir ff

**Review by Malax — CHANGES_REQUESTED** [review]
LGTM (not surprising since I've seen it a bunch of times before), the only thing that needs fixing before merge would be missing trailing newlines in several files.

Examples:
- buildpacks/php/tests/fixtures/platform/generator/base/expected_platform_composer.json
- buildpacks/php/tests/fixtures/platform/generator/composer1.10/expected_platform_composer.json
- .github/CODEOWNERS

**Review by Malax — APPROVED** [review]
🍾

File: buildpacks/php/src/platform.rs
Comment (dzuelke) [reply buildpacks/php/src/platform.rs:106]: Yeah, I'll think about better wording. It returns the struct for serialization by the layer for cleaner error handling, FWIW.

**Pull request merged**
